### ISSUE TITLE
reward problem

### ISSUE THREAD

**Issue by wanghua200** [reiniscimurs/DRL-robot-navigation#196]
Why is the reward when there is no collision or arrival calculated as action[0] * 0.5 - abs(action[1]) * 0.5 - r3(min_laser) / 2, with all terms divided by two?

**Comment by wanghua200** [comment]
Hello, thank you for sharing your work. I noticed that the paper doesn't show division by 2, but the code does. Do you think dividing by 2 is necessary? What benefits does it have?

**Comment by reiniscimurs** [comment]
Hi,

It lessens the influence of the immediate reward against the sparse reward. You do not have to divide by 2 but experimentally I found this to work better.

**Comment by wanghua200** [comment]
I really appreciate your answers. Are the learning rates for your actor and critic networks fixed, right?

**Comment by reiniscimurs** [comment]
Yes, it uses default learning rate, but you can expose it and set it in the optimizer:
https://github.com/reiniscimurs/DRL-robot-navigation/blob/017ed127928bb14532f15467498655f0c5b16600/TD3/train_velodyne_td3.py#L97

For example by setting:
```python
lr_actor=1e-4
lr_critic=3e-4
self.actor_optimizer = torch.optim.Adam(self.actor.parameters(), lr=lr_actor)
self.critic_optimizer = torch.optim.Adam(self.critic.parameters(), lr=lr_critic)

**Comment by wanghua200** [comment]
Hello! Thank you for your reply. What is the value range of the robot's linear velocity and angular velocity during training according to your code?

**Comment by wanghua200** [comment]
In the figure presented in your paper, the state input s is first passed through a fully connected layer (FC800) followed by a ReLU activation, and then transformed by a TFC600 layer to extract state features. The action input a is separately processed through another TFC600 layer to extract action features. These two branches are then fused by addition, followed by a ReLU activation. The fused features are further processed by a CFC600 layer before being passed through a final FC1 layer to output the Q-value. However, the structure in your code appears to differ from this design, as it does not clearly separate the state and action processing branches before fusion

**Comment by reiniscimurs** [comment]
This is given in the code:

https://github.com/reiniscimurs/DRL-robot-navigation/blob/017ed127928bb14532f15467498655f0c5b16600/TD3/train_velodyne_td3.py#L344-L345

**Comment by wanghua200** [comment]
In the code,
s11 = torch.mm(s1, self.layer_2_s.weight.data.t())
s12 = torch.mm(a, self.layer_2_a.weight.data.t())
s1 = F.relu(s11 + s12 + self.layer_2_a.bias.data)

The optimizer cannot update the parameters of these layers: self.layer_2_s and self.layer_2_a. Then, a large number of intermediate layers in the Critic network are not being trained at all. Is this reasonable?

**Comment by reiniscimurs** [comment]
This is an implementation of a version of a projection layer. Purpose of a projection layer is not to learn its weights, but rather to transform the incoming input into a different dimensionality representation.

https://stackoverflow.com/questions/37889914/what-is-a-projection-layer-in-the-context-of-neural-networks

**Comment by wanghua200** [comment]
Thank you for your reply. I gradually reduced the learning rate of the critic network during the middle stage of training, but there was no improvement. Do you know why?

**Comment by reiniscimurs** [comment]
Hi,

I do not know what you did in detail so I do not know what improvement was expected. Making the lr smaller will probably just increase the training time as the rate of update is made smaller but will not have effect on performance itself.

**Comment by wanghua200** [comment]
Thank you for your reply. In your article, I saw that you added delayed attribution rewards to guide navigation, but I didn't see this part in the code. Is there somewhere I missed?

**Comment by reiniscimurs** [comment]
It is not implemented in this repository. You can simply add a buffer of states before adding them to the replay buffer and reform the reward once a collision or goal is detected.

You will notice that this repository is created after the creation and acceptance of the paper. As such, it is a reimplementation of the specific training pipeline used in GDAE and is an example of how to train DRL navigation similarly to that used in the paper. But it uses a different API (pytorch instead of tensorflow) and has updated implementation with some changes.

**Comment by wanghua200** [comment]
I've noticed that other articles often include an immediate reward based on the difference in distance to the target between the current moment and the previous moment to guide navigation.However, in your code, the immediate reward is based on speed, angular velocity, and distance from obstacles. Do you think this is better? Thank you for your reply.

**Comment by wanghua200** [comment]
Could you please answer my question above? Thank you very much.

**Comment by reiniscimurs** [comment]
You can see some answers to that here: https://github.com/reiniscimurs/DRL-robot-navigation/issues/2#issuecomment-962382003

**Comment by wanghua200** [comment]
Thank you so much for your answers. In your code, an evaluation is performed every 5000 steps, and each evaluation consists of 10 episodes. However, many times an episode is less than 500 steps, so is it inappropriate to conduct an evaluation every 5000 steps? I think it would be more reasonable to do it every 10 episodes.

**Comment by reiniscimurs** [comment]
You can evaluate any way you want, it has no effect on training. This is entirely up to you.

But think of it this way. Lets say you evaluate every 10 episodes. But all of the episodes end in the first 5 steps. You have collected only 50 new samples during this time that the model is maybe trained on. Especially when episode lengths are small, you spend more time evaluating the model without any new information gain than actually training. But this is entirely up to you annd however you want to see the development of your model.

**Comment by wanghua200** [comment]
When training with distance as a dense reward, sometimes the robot starts to shake left and right without moving forward when it is very close to the target point. Do you know why? Thank you for your reply.

**Comment by reiniscimurs** [comment]
You can seach older discussions on this as it has been discussed to some extent. Policy is probably stuck in local optimum.

**Comment by wanghua200** [comment]
I noticed that you only initialize one critic network, but calculate the target and current Q-values twice, and then update the critic network using the sum of losses. What is the difference between this and initializing two critic networks? Does this conform to the original TD3?Thank you for your reply.

**Comment by reiniscimurs** [comment]
See the tutorial on the topic: https://medium.com/@reinis_86651/deep-reinforcement-learning-in-mobile-robot-navigation-tutorial-part2-network-82aea28b0e0f

**Comment by wanghua200** [comment]
I'm thinking of adjusting the frequency of policy updates based on the fluctuation of the loss. Do you think this is a good approach? And should I do this based on the loss fluctuation at each step or each episode? Thank you very much for your reply.

**Comment by reiniscimurs** [comment]
I think the time would be better spent looking into other optimizers and gradient clipping.

**Comment by wanghua200** [comment]
Thank you for your reply. Can I set the learning rate of the actor network to 0.0001 and the learning rate of the critic network to 0.001?

**Comment by reiniscimurs** [comment]
Sure. You can see how it is implemented here: https://github.com/reiniscimurs/DRL-robot-navigation-IR-SIM/blob/69815628cf07b145494dfec03df4c72b01759a68/robot_nav/models/TD3/TD3.py#L169-L175

**Comment by wanghua200** [comment]
Thank you for your reply.Sometimes, when training with the same parameters, the reward collision curves can vary significantly. How can I determine if there is an improvement after my optimization?

**Comment by reiniscimurs** [comment]
You mean between policies? 
You would need a test set or you can add your own KPIs. Feel free to use any that seem reasonable to you

**Comment by wanghua200** [comment]
Thank you for your reply. Is it necessary for me to focus on observing whether some curves in their training process, such as the reward curve and collision curve, are better? I find it very random. Without any changes, the convergence speed is not the same.

**Comment by reiniscimurs** [comment]
They are random because every episode is random, so that you will have different distances to goals in different situations. Rewards will fluctuate a lot over episodes. The convergence will also not be the same because every time you are training on different data due to different experiences and added noise. 

You can create a set of test scenarios and evaluate on those. That will give you an insight on how the performance is changing. Evaluating just on training KPIs will have a lot of variance without showing if the performance is better or not.

**Comment by wanghua200** [comment]
Thank you for your reply. I have noticed that when using linear velocity and angular velocity as immediate rewards, when the robot enters a corner and gets very close to obstacles, it will first only rotate and then move normally. In this case, doesn't the robot also only receive a negative reward when it is just rotating? Then, in that case, shouldn't the immediate reward calculation method based on distance also work? In your reply, you mentioned that with immediate rewards based on velocity and angular velocity, the robot will soon learn that the negative reward for rotating is smaller than that for colliding. I think, by the same logic, with immediate rewards based on distance, the reward for being far from the target point is also smaller than that for colliding. Could you please clarify this confusion for me? Thank you very much.

**Comment by reiniscimurs** [comment]
Sorry, I do not understand the setup of the distance reward here. My issue with distance to goal rewards is more about getting stuck in local optimum in some pockets near the goal, not that it wouldn't learn to avoid collisions. However, this also is highly dependent on how the reward structure is implemented and what the parameters for it are.

**Comment by wanghua200** [comment]
I recently want to add an LSTM to the network, but how to add an LSTM to the actor network? I have two ideas:

The first one:

Use a 400 - dimensional fully - connected layer to separately process the current state.
Use a 400 - dimensional LSTM layer to process the historical state sequence and take the output of the last time step.
Concatenate the processing result of the current state and the output of the LSTM.
Pass through fully - connected layers of 600 dimensions and 300 dimensions in sequence.
Finally, use the Tanh activation function to output the action, ensuring that the action value is within the range of [-1, 1].

The second one:

First, use an 800 - dimensional fully - connected layer to process the input state.
Pass the result into a 600 - dimensional LSTM layer to process the time - series information.
Take the output of the last time step of the LSTM and continue processing.
Pass through fully - connected layers of 600 dimensions and 300 dimensions in sequence.
Finally, use the Tanh activation function to output the action, ensuring that the action value is within the range of [-1, 1].

Which do you think is better? Thanks for your reply.

**Comment by wanghua200** [comment]
I want to solve the problem that the distance - based reward mechanism is prone to getting stuck in dead ends. I'm considering adding either GRU + Attention Mechanism or LSTM to the network. Which do you think is better? Thanks for your reply.

**Comment by reiniscimurs** [comment]
Theoretically LSTM should work better as it allows to "remember" states from further in the past while not influencing intermediate states. But you can try GRU/LSTM/RNN and experimentally find which one works best. They are easy to implement in pytorch and are easily interchangable. However, you will need to change the way how replay buffer stores and samples states.

You can see the implementation in:
https://github.com/reiniscimurs/DRL-robot-navigation-IR-SIM/tree/master/robot_nav/models/RCPG

**Comment by wanghua200** [comment]
In the environment and task you have set up, what do you think is a suitable sequence length? Would 25 be appropriate?Lines 142 and 143 of the code you provided:
self.layer_4 = nn.Linear(36, 400)
torch.nn.init.kaiming_uniform_(self.layer_1.weight, nonlinearity="leaky_relu")

Did you make a mistake here with self.layer_1.weight? Thank you for your help.

**Comment by reiniscimurs** [comment]
Hi,

Yes it should be `torch.nn.init.kaiming_uniform_(self.layer_4.weight, nonlinearity="leaky_relu")`

For sequence length, you can calculate for how many seconds do you want to mainain history and base your length on that. See what fits your methods.

You can also see an implementation already in this repo of RNNs here: https://github.com/Barry2333/DRL_Navigation

**Comment by wanghua200** [comment]
class Actor(nn.Module):
    def __init__(self, state_dim, action_dim):
        super(Actor, self).__init__()
        self.fc1 = nn.Linear(state_dim, 800)
        self.gru = nn.GRU(800, 600, batch_first=True)  
        self.fc2 = nn.Linear(600, 600)
        self.fc3 = nn.Linear(600, 300)
        self.output = nn.Linear(300, action_dim)
        self.tanh = nn.Tanh()
    def forward(self, state_seq):
        x = F.relu(self.fc1(state_seq))
        gru_out, _ = self.gru(x) 
        gru_last_out = gru_out[:, -1, :] 
        x = F.relu(self.fc2(gru_last_out))  
        x = F.relu(self.fc3(x))
        action = self.tanh(self.output(x))
        return action
class Critic(nn.Module):
    def __init__(self, state_dim, action_dim):
        super(Critic, self).__init__()
        # Q1 
        self.layer_1 = nn.Linear(state_dim, 800)
        self.gru_1 = nn.GRU(800, 600, batch_first=True)  
        self.layer_2_s = nn.Linear(600, 600)
        self.layer_2_a = nn.Linear(action_dim, 600)
        self.layer_3 = nn.Linear(600, 1)
        # Q2 
        self.layer_4 = nn.Linear(state_dim, 800)
        self.gru_2 = nn.GRU(800, 600, batch_first=True)  
        self.layer_5_s = nn.Linear(600, 600)
        self.layer_5_a = nn.Linear(action_dim, 600)
        self.layer_6 = nn.Linear(600, 1)

    def forward(self, state_seq, action):
        # state_seq: [B, T, state_dim]
        # action: [B, action_dim]
        # --- Q1 ---
        s1 = F.relu(self.layer_1(state_seq))  
        s1, _ = self.gru_1(s1)            
        s1 = s1[:, -1, :]                     
        self.layer_2_s(s1)  
        self.layer_2_a(action)
        s11 = torch.mm(s1, self.layer_2_s.weight.data.t())
        s12 = torch.mm(action, self.layer_2_a.weight.data.t())
        s1 = F.relu(s11 + s12 + self.layer_2_a.bias.data)
        q1 = self.layer_3(s1)
        # --- Q2 ---
        s2 = F.relu(self.layer_4(state_seq))
        s2, _ = self.gru_2(s2)  
        s2 = s2[:, -1, :]

        self.layer_5_s(s2)
        self.layer_5_a(action)

        s21 = torch.mm(s2, self.layer_5_s.weight.data.t())
        s22 = torch.mm(action, self.layer_5_a.weight.data.t())
        s2 = F.relu(s21 + s22 + self.layer_5_a.bias.data)
        q2 = self.layer_6(s2)

        return q1, q2  
 To better compare with the original version that doesn't use GRU, I've set up the network structure like this. Do you see any issues with it?  Thanks for your reply.

**Comment by reiniscimurs** [comment]
On a quick glance this seems fine to me, but in general, this is outside the scope of this repo and I can't really help much here. I encourage you to test it out and see if it performs the way you expect.

**Comment by wanghua200** [comment]
I noticed that when you defined the train method, you commented that the discount is 0.99, but when you set the global variable, you set it to 0.99999. In this case, the reward weights for the entire episode would be quite similar. From the other papers I've read, I found that they set it to 0.99 and 0.999. What was your consideration for setting it to 0.99999? I'm thinking of setting it to 0.999—do you think that's acceptable? Also, you set the batch size to 40, and I'm considering setting it to 128. Are these reasonable attempts? Thank you for your reply.

**Comment by reiniscimurs** [comment]
The discount factor defines how much of future revard you want to consider. The higher the factor the more estimated future reward influences the Q value.

Batch size and discount factors both are hyperparameters. You can tune and update them however you see fit for your specific usage.
